chips/sam4l/src/crccu.rs

```diff
@@ -0,0 +1,332 @@
+    { 0x1C, "DMA Interrupt Mask Register", DMAIMR, "R" },    // DMA interrupt enabled?
+    { 0x20, "DMA Interrupt Status Register", DMAISR, "R" },  // DMA transfer completed? (cleared when read)
+    { 0x34, "Control Register", CR, "W" },                   // Write a one to reset SR
+    { 0x38, "Mode Register", MR, "RW" },                     // Bandwidth divider, Polynomial type, Compare?, Enable?
+    { 0x3C, "Status Register", SR, "R" },                    // CRC result (unreadable if MR.COMPARE=1)
+    { 0x40, "Interrupt Enable Register", IER, "W" },         // Write one to set IMR.ERR bit (zero no effect)
+    { 0x44, "Interrupt Disable Register", IDR, "W" },        // Write zero to clear IMR.ERR bit (one no effect)
+    { 0x48, "Interrupt Mask Register", IMR, "R" },           // If IMR.ERR bit is set, error-interrupt (for compare) is enabled
+    { 0x4C, "Interrupt Status Register", ISR, "R" },         // CRC error (for compare)? (cleared when read)
+    { 0xFC, "Version Register", VERSION, "R" }               // 12 low-order bits: version of this module.  = 0x00000202
+];
+
+// CRCCU Descriptor (from Table 41.2 in Section 41.6):
+#[repr(C, packed)]
+struct Descriptor {
+    addr: u32,       // Transfer Address Register (RW): Address of memory block to compute
+    ctrl: TCR,       // Transfer Control Register (RW): IEN, TRWIDTH, BTSIZE
+    _res: [u32; 2],
+    crc: u32         // Transfer Reference Register (RW): Reference CRC (for compare mode)
+}
+
+// Transfer Control Register (see Section 41.6.18)
+#[derive(Copy, Clone)]
+#[repr(C, packed)]
+struct TCR(u32);
+
+impl TCR {
+    const fn new(enable_interrupt: bool, trwidth: TrWidth, btsize: u16) -> Self {
+        TCR((!enable_interrupt as u32) << 27
+            | (trwidth as u32) << 24
+            | (btsize as u32))
+    }
+
+    const fn default() -> Self {
+        Self::new(false, TrWidth::Byte, 0)
+    }
+
+    fn interrupt_enabled(self) -> bool {
+        (self.0 & (1 << 27)) == 0
+    }
+
+    #[allow(dead_code)]
+    fn get_btsize(self) -> u16 {
+        (self.0 & 0xffff) as u16
+    }
+}
+
+pub enum TrWidth { Byte, HalfWord, Word }
+
+// Mode Register (see Section 41.6.10)
+struct Mode(u32);
+
+impl Mode {
+	fn new(divider: u8, ptype: Polynomial, compare: bool, enable: bool) -> Self {
+        Mode((((divider & 0x0f) as u32) << 4)
+             | (ptype as u32) << 2
+             | (compare as u32) << 1
+             | (enable as u32))
+    }
+    fn disabled() -> Self {
+        Mode::new(0, Polynomial::CCIT8023, false, false)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum State { Invalid, Initialized, Enabled }
+
+// State for managing the CRCCU
+pub struct Crccu<'a> {
+    client: Option<&'a crc::Client>,
+    state: Cell<State>,
+
+    // Guaranteed room for a Descriptor with 512-byte alignment.
+    // (Can we do this statically instead?)
+    descriptor_space: [u8; DSCR_RESERVE],
```

Comment:
I think there is safer way of doing all of this, where you explicitly model the descriptor type, including the reserved space:

```rust
struct Descriptor {
    addr: u32,       // Transfer Address Register (RW): Address of memory block to compute
    ctrl: TCR,       // Transfer Control Register (RW): IEN, TRWIDTH, BTSIZE
    _res: [u32; 2],
    crc: u32         // Transfer Reference Register (RW): Reference CRC (for compare mode)
    space: [u8; DSCR_RESERVE]
}

...

pub struct Crccu<'a> {
    client: Option<&'a crc::Client>,
    state: Cell<State>,
    descriptor: TakeCell<'static, Descriptor>
}
```

The only thing to worry about is allocating this `Descriptor` but that can either be just statically allocated or passed into `new` (in which case the board will statically allocate it).